File: Python/200. Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if len(grid) == 0 or len(grid[0]) == 0:
            return 0
        m, n = len(grid), len(grid[0])        
        ans = 0
        def dfs(i, j):
            grid[i][j] = '0'
            dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]
            for d in dirs:
                x, y = i + d[0], j + d[1]
                if x < 0 or x >= m or y < 0 or y >=n or grid[x][y] == '0':
                    continue
                dfs(x, y)            
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':                    
                    ans += 1
                    dfs(i, j)
        return ans
                    

File: 2024/200. Number of Islands.cpp
class Solution {
public:
    int dirs[4][2] = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    int numIslands(vector<vector<char>>& grid) {
        int count = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    grid[i][j] = '0';
                    count += 1;
                    dfs(grid, i, j);
                }
            }
        }
        return count;
    }

    void dfs(vector<vector<char>>& grid, int x, int y) {
        for (int i = 0; i < 4; i++) {
            int nx = x + dirs[i][0], ny = y + dirs[i][1];
            if (nx < 0 || nx >= grid.size() || ny < 0 || ny >= grid[0].size() ||
                grid[nx][ny] == '0') {
                continue;
            }
            grid[nx][ny] = '0';
            dfs(grid, nx, ny);
        }
    }
};